monitor/main.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
+
+from sets import Set
+
+import logging
+import os
+import time
+
+from prometheus_client import start_http_server, Gauge
+from kubernetes import client, config
+from kubernetes.client import ApiClient, Configuration
+from openshift.dynamic import DynamicClient
+
+MACHINE_STATUS = Gauge('machine_api_status',"Status of Machine CRs", labelnames=['machine_name','node_name','instance_state','instance_id'])
+
+# A list (implemented as a Set) of all active Machines
+ACTIVE_MACHINES = Set([])
+
+def get_machines(dynamic_client,namespace):
+    """Gets all of the Machine objects from the cluster from the specified namespace
+    """
+    machines = dynamic_client.resources.get(kind='Machine')
+    return machines.get(namespace=namespace).items
+
+def collect(dynamic_client, namespace):
+    """
+    Collect the current data from the AWS API
+    """
+
+    # List of volumes that we've actually had data back for the API
+    seen_machines = Set([])
+    machines = get_machines(dynamic_client, namespace)
+    for machine in machines:
+        seen_machines.add(machine['metadata']['name'])
+        ACTIVE_MACHINES.add(machine['metadata']['name'])
+
+        node_name = ""
+        instanceid = ""
+        value = 1
+        if 'nodeRef' in machine['status'].keys():
+            node_name = machine['status']['nodeRef']['name']
+        else:
+            value = 0
+
+        if 'instanceId' in machine['status']['providerStatus'].keys():
+            instanceid = machine['status']['providerStatus']['instanceId']
+        else:
+            value = 0
+        MACHINE_STATUS.labels(
+            machine_name   = machine['metadata']['name'],
+            node_name      = node_name,
+            instance_state = machine['status']['providerStatus']['instanceState'],
+            instance_id    = instanceid
+            ).set(value)
+
+    logging.debug("Have %d ACTIVE_MACHINES, seen %d machines, total machines from list_metrics %d",len(ACTIVE_MACHINES),len(seen_machines),len(machines))
+    for inactive_machine in ACTIVE_MACHINES - seen_machines:
+        logging.info("Removing machine_id='%s' from Prometheus ",inactive_machine)
+        node_name = ""
+        instanceid = ""
+        if 'nodeRef' in machine['status'].keys():
+            node_name = machine['status']['nodeRef']['name']
+        if 'instanceId' in machine['status']['providerStatus'].keys():
+            instanceid = machine['status']['providerStatus']['instanceId']
+
+        MACHINE_STATUS.remove(
+            machine_name   = inactive_machine['metadata']['name'],
+            node_name      = node_name,
+            instance_state = inactive_machine['status']['providerStatus']['instanceState'],
```

Comment:
Key `instanceState` won't exist if the machine has no instance.  Need to check it like `instanceId` and `nodeRef`.

```
...
status:
  lastUpdated: 2019-06-19T20:37:09Z
  providerStatus:
    apiVersion: awsproviderconfig.openshift.io/v1beta1
    conditions:
    - lastProbeTime: 2019-06-19T20:37:09Z
      lastTransitionTime: 2019-06-17T10:01:21Z
      message: "error launching instance: error creating EC2 instance: InstanceLimitExceeded:
        You have requested more instances (7) than your current instance limit of
        5 allows for the specified instance type. Please visit http://aws.amazon.com/contact-us/ec2-request
        to request an adjustment to this limit.\n\tstatus code: 400, request id: 2685529d-ad0c-4658-aac7-a91a59df1b49"
      reason: MachineCreationFailed
      status: "True"
      type: MachineCreation
    kind: AWSMachineProviderStatus
```

Reply:
This works out to:

```srep_1539_machine_status{instance_id="",instance_state="None",machine_name="lseelye-test9-pjqm5-worker-ca-central-1a-6jdk4",node_name=""} 0.0
```

`instance_state` is "`None`" and I think that's reasonable?

Reply:
If `None` is ok for this why not for the other fields?  Seems we could just do `node_name` and `instance_id` in-line here as well and have labels with `None` value for consistency.

Reply:
@jewzaam I'm hesitant to have that be a "magic value" for keys that have arbitrary values (that is, the `instance_id` and `node_name`). In contrast, "None" can be a new state that we can be reasonably sure will not clobber an existing one from the operator. Maybe we want the state to be a different value we can be equally certain will not clobber a pre-existing one?

Reply:
Based on the CRD:

```
  - JSONPath: .status.providerStatus.instanceState
    description: State of the AWS instance
    name: State  
    type: string 
```

This sounds like a passthrough from AWS.  In a cluster I'm looking at currently the values are all `running`.

Valid states per AWS docs: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html

- pending
- running
- stopping
- stopped
- shutting-down
- terminated

To be consistent, could we make it lower case?

Reply:
Can do...

If we're going to downcase it we might as well create a different string since we'll already need to interact with the concept of `None` in Python (eg casting it to a string).

Some options: `absent`, `none`, `not-created`, `missing`, `gone`. I slightly prefer `not-created` but have no strong preference. I'll update the PR for that and we can bikeshed about the name later :)